doc/neps/nep-0052-python-api-cleanup.rst

```diff
@@ -0,0 +1,333 @@
+  `gh-12385 <https://github.com/numpy/numpy/issues/12385>`__
+  can be deprecated and removed.
+- All ``*sctype`` functions can be deprecated and removed, they (see
+  `gh-17325 <https://github.com/numpy/numpy/issues/17325>`__,
+  `gh-12334 <https://github.com/numpy/numpy/issues/12334>`__,
+  and other issues for ``maximum_sctype`` and related functions).
+- Business day functionality can likely be removed (unclear if it needs
+  splitting out like was done for ``np.financial``).
+- The ``np.compat`` namespace, used during the Python 2 to 3 transition, will be removed.
+- Functions that are narrow in scope, with very few public use-cases,
+  will be removed.  See, e.g.
+  ``real_if_close`` (`gh-11375 <https://github.com/numpy/numpy/issues/11375>`__).
+  These will have to be identified manually and by issue triage.
+
+New namespaces are introduced for warnings/exceptions (``np.exceptions``) and
+for dtype-related functionality (``np.types``). NumPy 2.0 is a good opportunity
+to populate these submodules from the main namespace.
+
+Functionality that is widely used but has a preferred alternative may either be
+deprecated (with the deprecation message pointing out what to use instead) or
+be hidden by not including it in ``__dir__``. In case of hiding, a ``..
+legacy::`` directory may be used to mark such functionality in the
+documentation.
+
+A test will be added to ensure limited future growth of all namespaces; i.e.,
+every new entry will need to be explicitly added to an allow-list.
+
+
+Cleaning up the submodule structure
+```````````````````````````````````
+
+We will clean up the NumPy submodule structure, so it is easier to navigate.
+When this was discussed before (see
+`MAINT: Hide internals of np.lib to only show submodules <https://github.com/numpy/numpy/pull/18447>`__)
+there was already rough consensus on that - however it was hard to pull off in
+a minor release.
+
+A basic principle we will adhere to is "one function, one location". Functions
+that are exposed in more than one namespace (e.g., many functions are present
+in ``numpy`` and ``numpy.lib``) need to find a single home.
+
+We will reorganize the API reference guide along main and submodule namespaces,
+and only within the main namespace use the current subdivision along
+functionality groupings. Also by "mainstream" and special-purpose namespaces:
+
+::
+
+    # Regular/recommended user-facing namespaces for general use. Present these
+    # as the primary set of namespaces to the users.
+    numpy
+    numpy.exceptions
+    numpy.fft
+    numpy.linalg
+    numpy.polynomial
+    numpy.random
+    numpy.testing
+    numpy.typing
+
+    # Special-purpose namespaces. Keep these, but document them in a separate
+    # grouping in the reference guide and explain their purpose.
+    numpy.array_api
+    numpy.ctypeslib
+    numpy.emath
+    numpy.f2py  # only a couple of public functions, like `compile` and `get_include`
+    numpy.lib.stride_tricks
```

Comment:
What about moving `stride_tricks` to `numpy.stride_tricks`, and considering the rest of `numpy.lib` legacy stuff?

Reply:
That's a reasonable option; there's a few more public things in `numpy.lib` though:

- [lib.format](https://numpy.org/devdocs/reference/generated/numpy.lib.format.html#module-numpy.lib.format) (module with multiple functions)
- [lib.mixins.NDArrayOperatorsMixin](https://numpy.org/devdocs/reference/generated/numpy.lib.mixins.NDArrayOperatorsMixin.html#numpy.lib.mixins.NDArrayOperatorsMixin) (module with a single object)
- [lib.NumPyVersion](https://numpy.org/devdocs/reference/generated/numpy.lib.NumpyVersion.html)
- [lib.Arrayterator](https://numpy.org/devdocs/reference/generated/numpy.lib.Arrayterator.html#numpy-lib-arrayterator)

I think that's all, but hard to be 100% sure. It's hard to find good places for each of them, so I had in mind to keep `np.lib` as a grab bag of stuff that does not deserve its own namespace right below the main namespace.

We can probably get rid of `Arrayterator`; I'm not sure about the rest.

Reply:
I've added the list above to the note. I'd like to resolve this later (post-merge of this NEP as Draft). Leaning towards keeping ``numpy.lib``, but either way is not ideal.